### ISSUE TITLE
👊 👊 🍖

### ISSUE THREAD

**Issue by whit537** [gratipay/gratipay.com#4615]
Trying to get commits from https://github.com/gratipay/gratipay.com/pull/4613 into some kind of usable review pipeline. I'm in a mess with Braintree remote API creds during testing and VCR (re)recording.

**Comment by whit537** [comment]
We have two sets of creds, one in `defaults.env`, the other in `tests/defaults.env` (added in 6d8b0a14086348c260d1690739295f0654a8f6c1, I think just for convenience).

**Comment by whit537** [comment]
VCR cassettes intermingle Braintree API calls with all of the other API calls, so if we want to rerecord all of the Braintree APIs then its work to do that without rerecording the other API calls. I mean, we rerecord whenever we CI master at Travis and that appears to be green so it should work, right?

**Comment by whit537** [comment]
The problem I'm hitting is that the account represented by the creds currently in `tests/defaults.env` doesn't have the custom field I need in order to record these new API calls against it.

**Comment by whit537** [comment]
Why does it not work to rerecord existing tests against a new Braintree account?

**Comment by whit537** [comment]
Oooh! Was there some special casing for `BillingHarness`? A need to run that in a special way by itself or something?

**Comment by whit537** [comment]
```
$ gs
 D tests/py/fixtures/BillingHarness.yml
 D tests/py/fixtures/ConsumeChangeStreamTests.yml
 D tests/py/fixtures/TestElsewhere.yml
 D tests/py/fixtures/TestFriendFinder.yml
 D tests/py/fixtures/TestHistory.yml
 D tests/py/fixtures/TestNMassPays.yml
 D tests/py/fixtures/TestNewTeams.yml
 D tests/py/fixtures/TestPages.yml
 D tests/py/fixtures/TestPayday.yml
 D tests/py/fixtures/TestPayin.yml
 D tests/py/fixtures/TestRoutes.yml
 D tests/py/fixtures/TestTeamEdit.yml
 D tests/py/fixtures/TestTeams.yml
 D tests/py/fixtures/TestsWithBillingHarness.yml
$
```

**Comment by whit537** [comment]
Ah! Need the old `custom_field`, ya?

**Comment by whit537** [comment]
Appears that way ...

```
(Pdb) result
<ErrorResult 'Custom field is invalid: participant_id.' at 10ce54890>
(Pdb)
```

... but ...

![screen shot 2017-09-09 at 8 20 25 pm](https://user-images.githubusercontent.com/134455/30245063-5391044c-959c-11e7-89e6-95052091492c.png)

🤔

**Comment by whit537** [comment]
Am I hitting the right account?

**Comment by whit537** [comment]
I went to check production to see how it's set there. I guess I typed the password wrong because I got "Your account is now locked out. Please wait 30 minutes and try again." For one bad password attempt? 😳

**Comment by whit537** [comment]
False alarm! `gratipay` vs. `gratipay_`

**Comment by whit537** [comment]
# Production

![screen shot 2017-09-09 at 8 24 08 pm](https://user-images.githubusercontent.com/134455/30245079-dad99c52-959c-11e7-84a3-36d47118b91c.png)

**Comment by whit537** [comment]
# Sandbox

![screen shot 2017-09-09 at 8 24 34 pm](https://user-images.githubusercontent.com/134455/30245083-eacbfaa6-959c-11e7-8079-726b37af345b.png)

**Comment by whit537** [comment]
Oh! Need to re-record. 😊

**Comment by whit537** [comment]
```
(Pdb) result
<SuccessfulResult {customer: <Customer {first_name: None, last_name: None, id: u'[]'} at []>} at []>
(Pdb)
```

💃

**Comment by whit537** [comment]
Okay! How many tests does that fix? 😁

**Comment by whit537** [comment]
Gah! Now hitting 'Gateway Rejected: duplicate', something I saw on #4613 and reported to Braintree support—don't expect that from sandbox requests.

**Comment by whit537** [comment]
Workaround over there is [purge sandbox](https://developers.braintreepayments.com/reference/general/testing/python#purge-sandbox-data) (takes a few minutes).

**Comment by whit537** [comment]
(And, helpfully, *doesn't* whack custom fields. 👍 )

**Comment by whit537** [comment]
(I've logged out of production. Better not go back in there tonight except hopefully to set the custom field!)

**Comment by whit537** [comment]
Aaaaaand of course my wireless interface died [again](https://github.com/gratipay/inside.gratipay.com/issues/1149#issuecomment-326522094). 🐭  Happened a couple days ago, too. Getting worse. Had to reboot several times and actually power down and back up. Skin of our teeth, @Gratipay!

**Comment by whit537** [comment]
Yeah, okay, back to these duplicates. :-/

I guess let's purge the sandbox and try again?

**Comment by whit537** [comment]
Okay, I'm back in. Rerunning ...

**Comment by whit537** [comment]
Skunked.

```
================================================== FAILURES ==================================================
_______________________________ TestsWithBillingHarness.test_capch_full_amount _______________________________
Traceback (most recent call last):
  File "/Users/whit537/personal/gratipay/gratipay.com/tests/py/test_billing_exchanges.py", line 115, in test_capch_full_amount
    assert error == ''  # sanity check
AssertionError: assert 'Gateway Rejected: duplicate' == ''
  - Gateway Rejected: duplicate
-------------------------------------------- Captured stdout call --------------------------------------------
pid-768 thread-140736820274112 (MainThread) Holding 2091 cents ($20.00 + $0.91 fee = $20.91) on Braintree for obama ... failed: Gateway Rejected: duplicate
_____________________________ TestsWithBillingHarness.test_capch_partial_amount ______________________________
Traceback (most recent call last):
  File "/Users/whit537/personal/gratipay/gratipay.com/tests/py/test_billing_exchanges.py", line 126, in test_capch_partial_amount
    assert error == ''  # sanity check
AssertionError: assert 'Gateway Rejected: duplicate' == ''
  - Gateway Rejected: duplicate
-------------------------------------------- Captured stdout call --------------------------------------------
pid-768 thread-140736820274112 (MainThread) Holding 2091 cents ($20.00 + $0.91 fee = $20.91) on Braintree for obama ... failed: Gateway Rejected: duplicate
_____________________________ TestsWithBillingHarness.test_capch_too_high_amount _____________________________
Traceback (most recent call last):
  File "/Users/whit537/personal/gratipay/gratipay.com/tests/py/test_billing_exchanges.py", line 134, in test_capch_too_high_amount
    assert error == ''  # sanity check
AssertionError: assert 'Gateway Rejected: duplicate' == ''
  - Gateway Rejected: duplicate
-------------------------------------------- Captured stdout call --------------------------------------------
pid-768 thread-140736820274112 (MainThread) Holding 2091 cents ($20.00 + $0.91 fee = $20.91) on Braintree for obama ... failed: Gateway Rejected: duplicate
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! KeyboardInterrupt !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
/usr/local/Cellar/python/2.7.13_1/Frameworks/Python.framework/Versions/2.7/lib/python2.7/ssl.py:653: KeyboardInterrupt
```

**Comment by whit537** [comment]
YES!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! 💃 

![screen shot 2017-09-09 at 8 55 29 pm](https://user-images.githubusercontent.com/134455/30245254-3eceea1a-95a1-11e7-821d-f2d3ae84e2ab.png)

**Comment by whit537** [comment]
```
tests/py/fixtures/BillingHarness.yml
6:    uri: https://api.sandbox.braintreegateway.com:443/merchants/bk8h97tqzyqjhtfn/customers
28:    uri: https://api.sandbox.braintreegateway.com:443/merchants/bk8h97tqzyqjhtfn/customers
50:    uri: https://api.sandbox.braintreegateway.com:443/merchants/bk8h97tqzyqjhtfn/payment_methods

tests/py/fixtures/TestsWithBillingHarness.yml
8:    uri: https://api.sandbox.braintreegateway.com:443/merchants/bk8h97tqzyqjhtfn/transactions
50:    uri: https://api.sandbox.braintreegateway.com:443/merchants/bk8h97tqzyqjhtfn/transactions/awtzjbc6/submit_for_settlement
93:    uri: https://api.sandbox.braintreegateway.com:443/merchants/bk8h97tqzyqjhtfn/transactions/awtzjbc6/void
138:    uri: https://api.sandbox.braintreegateway.com:443/merchants/bk8h97tqzyqjhtfn/transactions
180:    uri: https://api.sandbox.braintreegateway.com:443/merchants/bk8h97tqzyqjhtfn/transactions/fpjm0ep0/submit_for_settlement
222:    uri: https://api.sandbox.braintreegateway.com:443/merchants/bk8h97tqzyqjhtfn/transactions/fpjm0ep0
264:    uri: https://api.sandbox.braintreegateway.com:443/merchants/bk8h97tqzyqjhtfn/transactions/fpjm0ep0/void
309:    uri: https://api.sandbox.braintreegateway.com:443/merchants/bk8h97tqzyqjhtfn/transactions
351:    uri: https://api.sandbox.braintreegateway.com:443/merchants/bk8h97tqzyqjhtfn/transactions/59fd1wvd/submit_for_settlement
394:    uri: https://api.sandbox.braintreegateway.com:443/merchants/bk8h97tqzyqjhtfn/transactions/59fd1wvd/void
439:    uri: https://api.sandbox.braintreegateway.com:443/merchants/bk8h97tqzyqjhtfn/transactions
481:    uri: https://api.sandbox.braintreegateway.com:443/merchants/bk8h97tqzyqjhtfn/transactions/a371g057/submit_for_settlement
525:    uri: https://api.sandbox.braintreegateway.com:443/merchants/bk8h97tqzyqjhtfn/transactions/a371g057/void
567:    uri: https://api.sandbox.braintreegateway.com:443/merchants/bk8h97tqzyqjhtfn/customers
589:    uri: https://api.sandbox.braintreegateway.com:443/merchants/bk8h97tqzyqjhtfn/payment_methods
621:    uri: https://api.sandbox.braintreegateway.com:443/merchants/bk8h97tqzyqjhtfn/transactions
664:    uri: https://api.sandbox.braintreegateway.com:443/merchants/bk8h97tqzyqjhtfn/customers
686:    uri: https://api.sandbox.braintreegateway.com:443/merchants/bk8h97tqzyqjhtfn/payment_methods
716:    uri: https://api.sandbox.braintreegateway.com:443/merchants/bk8h97tqzyqjhtfn/payment_methods
748:    uri: https://api.sandbox.braintreegateway.com:443/merchants/bk8h97tqzyqjhtfn/transactions
790:    uri: https://api.sandbox.braintreegateway.com:443/merchants/bk8h97tqzyqjhtfn/transactions/res21cek/void
834:    uri: https://api.sandbox.braintreegateway.com:443/merchants/bk8h97tqzyqjhtfn/transactions
876:    uri: https://api.sandbox.braintreegateway.com:443/merchants/bk8h97tqzyqjhtfn/transactions/7mpag0w8/void
920:    uri: https://api.sandbox.braintreegateway.com:443/merchants/bk8h97tqzyqjhtfn/transactions
962:    uri: https://api.sandbox.braintreegateway.com:443/merchants/bk8h97tqzyqjhtfn/transactions/3mjm48sd/submit_for_settlement
1006:    uri: https://api.sandbox.braintreegateway.com:443/merchants/bk8h97tqzyqjhtfn/transactions
1048:    uri: https://api.sandbox.braintreegateway.com:443/merchants/bk8h97tqzyqjhtfn/transactions/h85kc7fw/submit_for_settlement
1093:    uri: https://api.sandbox.braintreegateway.com:443/merchants/bk8h97tqzyqjhtfn/transactions
1135:    uri: https://api.sandbox.braintreegateway.com:443/merchants/bk8h97tqzyqjhtfn/transactions/d81x0fd1/submit_for_settlement
1180:    uri: https://api.sandbox.braintreegateway.com:443/merchants/bk8h97tqzyqjhtfn/transactions
1222:    uri: https://api.sandbox.braintreegateway.com:443/merchants/bk8h97tqzyqjhtfn/transactions/2typ1xje/submit_for_settlement
1265:    uri: https://api.sandbox.braintreegateway.com:443/merchants/bk8h97tqzyqjhtfn/transactions/advanced_search_ids
```

**Comment by whit537** [comment]
Oh. 😁 

```
$ gs
 M tests/py/fixtures/BillingHarness.yml
 D tests/py/fixtures/ConsumeChangeStreamTests.yml
 D tests/py/fixtures/TestElsewhere.yml
 D tests/py/fixtures/TestFriendFinder.yml
 D tests/py/fixtures/TestHistory.yml
 D tests/py/fixtures/TestNMassPays.yml
 D tests/py/fixtures/TestNewTeams.yml
 D tests/py/fixtures/TestPages.yml
 D tests/py/fixtures/TestPayday.yml
 D tests/py/fixtures/TestPayin.yml
 D tests/py/fixtures/TestRoutes.yml
 D tests/py/fixtures/TestTeamEdit.yml
 D tests/py/fixtures/TestTeams.yml
 M tests/py/fixtures/TestsWithBillingHarness.yml
$
```

**Comment by whit537** [comment]
```
tests/py/fixtures/BillingHarness.yml
6:    uri: https://api.sandbox.braintreegateway.com:443/merchants/j9gwdfjdkxymhdgr/customers
28:    uri: https://api.sandbox.braintreegateway.com:443/merchants/j9gwdfjdkxymhdgr/customers
50:    uri: https://api.sandbox.braintreegateway.com:443/merchants/j9gwdfjdkxymhdgr/payment_methods

tests/py/fixtures/TestHistory.yml
8:    uri: https://api.sandbox.braintreegateway.com:443/merchants/j9gwdfjdkxymhdgr/transactions
49:    uri: https://api.sandbox.braintreegateway.com:443/merchants/j9gwdfjdkxymhdgr/transactions/6tvkry/submit_for_settlement
93:    uri: https://api.sandbox.braintreegateway.com:443/merchants/j9gwdfjdkxymhdgr/transactions
134:    uri: https://api.sandbox.braintreegateway.com:443/merchants/j9gwdfjdkxymhdgr/transactions/4w92rd/submit_for_settlement
176:    uri: https://api.sandbox.braintreegateway.com:443/merchants/j9gwdfjdkxymhdgr/transactions/advanced_search_ids

tests/py/fixtures/TestNMassPays.yml
8:    uri: https://api.sandbox.braintreegateway.com:443/merchants/j9gwdfjdkxymhdgr/transactions
50:    uri: https://api.sandbox.braintreegateway.com:443/merchants/j9gwdfjdkxymhdgr/transactions/k0tqp8ks/submit_for_settlement
94:    uri: https://api.sandbox.braintreegateway.com:443/merchants/j9gwdfjdkxymhdgr/transactions
136:    uri: https://api.sandbox.braintreegateway.com:443/merchants/j9gwdfjdkxymhdgr/transactions/e733859z/submit_for_settlement
180:    uri: https://api.sandbox.braintreegateway.com:443/merchants/j9gwdfjdkxymhdgr/transactions
222:    uri: https://api.sandbox.braintreegateway.com:443/merchants/j9gwdfjdkxymhdgr/transactions/1dygf0mg/submit_for_settlement
264:    uri: https://api.sandbox.braintreegateway.com:443/merchants/j9gwdfjdkxymhdgr/transactions/advanced_search_ids

tests/py/fixtures/TestPages.yml
6:    uri: https://api.sandbox.braintreegateway.com:443/merchants/j9gwdfjdkxymhdgr/customers
28:    uri: https://api.sandbox.braintreegateway.com:443/merchants/j9gwdfjdkxymhdgr/payment_methods
128:    uri: https://api.sandbox.braintreegateway.com:443/merchants/j9gwdfjdkxymhdgr/customers
150:    uri: https://api.sandbox.braintreegateway.com:443/merchants/j9gwdfjdkxymhdgr/payment_methods
250:    uri: https://api.sandbox.braintreegateway.com:443/merchants/j9gwdfjdkxymhdgr/payment_methods/any/cj4ztr
280:    uri: https://api.sandbox.braintreegateway.com:443/merchants/j9gwdfjdkxymhdgr/payment_methods/any/cj4ztr
310:    uri: https://api.sandbox.braintreegateway.com:443/merchants/j9gwdfjdkxymhdgr/customers/33477840
344:    uri: https://api.sandbox.braintreegateway.com:443/merchants/j9gwdfjdkxymhdgr/client_token
380:    uri: https://api.sandbox.braintreegateway.com:443/merchants/j9gwdfjdkxymhdgr/customers
402:    uri: https://api.sandbox.braintreegateway.com:443/merchants/j9gwdfjdkxymhdgr/payment_methods
502:    uri: https://api.sandbox.braintreegateway.com:443/merchants/j9gwdfjdkxymhdgr/payment_methods/any/gxny4r
532:    uri: https://api.sandbox.braintreegateway.com:443/merchants/j9gwdfjdkxymhdgr/payment_methods/any/gxny4r
562:    uri: https://api.sandbox.braintreegateway.com:443/merchants/j9gwdfjdkxymhdgr/customers/12743055
596:    uri: https://api.sandbox.braintreegateway.com:443/merchants/j9gwdfjdkxymhdgr/client_token

tests/py/fixtures/TestPayday.yml
8:    uri: https://api.sandbox.braintreegateway.com:443/merchants/j9gwdfjdkxymhdgr/transactions
49:    uri: https://api.sandbox.braintreegateway.com:443/merchants/j9gwdfjdkxymhdgr/transactions/c9xhq4/submit_for_settlement
93:    uri: https://api.sandbox.braintreegateway.com:443/merchants/j9gwdfjdkxymhdgr/transactions
134:    uri: https://api.sandbox.braintreegateway.com:443/merchants/j9gwdfjdkxymhdgr/transactions/k5q4rt/submit_for_settlement
178:    uri: https://api.sandbox.braintreegateway.com:443/merchants/j9gwdfjdkxymhdgr/transactions
219:    uri: https://api.sandbox.braintreegateway.com:443/merchants/j9gwdfjdkxymhdgr/transactions/d2dqbt/submit_for_settlement
263:    uri: https://api.sandbox.braintreegateway.com:443/merchants/j9gwdfjdkxymhdgr/transactions
304:    uri: https://api.sandbox.braintreegateway.com:443/merchants/j9gwdfjdkxymhdgr/transactions/38r9r8/submit_for_settlement
346:    uri: https://api.sandbox.braintreegateway.com:443/merchants/j9gwdfjdkxymhdgr/transactions/advanced_search_ids

tests/py/fixtures/TestPayin.yml
6:    uri: https://api.sandbox.braintreegateway.com:443/merchants/j9gwdfjdkxymhdgr/transactions/advanced_search_ids
27:    uri: https://api.sandbox.braintreegateway.com:443/merchants/j9gwdfjdkxymhdgr/transactions
68:    uri: https://api.sandbox.braintreegateway.com:443/merchants/j9gwdfjdkxymhdgr/transactions/bthrmy/void
110:    uri: https://api.sandbox.braintreegateway.com:443/merchants/j9gwdfjdkxymhdgr/transactions/bthrmy
154:    uri: https://api.sandbox.braintreegateway.com:443/merchants/j9gwdfjdkxymhdgr/transactions
195:    uri: https://api.sandbox.braintreegateway.com:443/merchants/j9gwdfjdkxymhdgr/transactions/advanced_search_ids
216:    uri: https://api.sandbox.braintreegateway.com:443/merchants/j9gwdfjdkxymhdgr/transactions/advanced_search
259:    uri: https://api.sandbox.braintreegateway.com:443/merchants/j9gwdfjdkxymhdgr/transactions/advanced_search_ids
280:    uri: https://api.sandbox.braintreegateway.com:443/merchants/j9gwdfjdkxymhdgr/transactions/advanced_search
323:    uri: https://api.sandbox.braintreegateway.com:443/merchants/j9gwdfjdkxymhdgr/transactions/8vx7md/void

tests/py/fixtures/TestRoutes.yml
7:    uri: https://api.sandbox.braintreegateway.com:443/merchants/j9gwdfjdkxymhdgr/payment_methods
42:    uri: https://api.sandbox.braintreegateway.com:443/merchants/j9gwdfjdkxymhdgr/customers/35060504
81:    uri: https://api.sandbox.braintreegateway.com:443/merchants/j9gwdfjdkxymhdgr/payment_methods/any/9yg226
99:    uri: https://api.sandbox.braintreegateway.com:443/merchants/j9gwdfjdkxymhdgr/customers/35060504
125:    uri: https://api.sandbox.braintreegateway.com:443/merchants/j9gwdfjdkxymhdgr/payment_methods
148:    uri: https://api.sandbox.braintreegateway.com:443/merchants/j9gwdfjdkxymhdgr/payment_methods/any/bkhpqm
178:    uri: https://api.sandbox.braintreegateway.com:443/merchants/j9gwdfjdkxymhdgr/customers/11443524
212:    uri: https://api.sandbox.braintreegateway.com:443/merchants/j9gwdfjdkxymhdgr/client_token
248:    uri: https://api.sandbox.braintreegateway.com:443/merchants/j9gwdfjdkxymhdgr/customers
271:    uri: https://api.sandbox.braintreegateway.com:443/merchants/j9gwdfjdkxymhdgr/client_token
307:    uri: https://api.sandbox.braintreegateway.com:443/merchants/j9gwdfjdkxymhdgr/payment_methods/any/bkhpqm
337:    uri: https://api.sandbox.braintreegateway.com:443/merchants/j9gwdfjdkxymhdgr/customers/11443524
371:    uri: https://api.sandbox.braintreegateway.com:443/merchants/j9gwdfjdkxymhdgr/client_token
407:    uri: https://api.sandbox.braintreegateway.com:443/merchants/j9gwdfjdkxymhdgr/payment_methods/any/bkhpqm
437:    uri: https://api.sandbox.braintreegateway.com:443/merchants/j9gwdfjdkxymhdgr/transactions/advanced_search_ids

tests/py/fixtures/TestsWithBillingHarness.yml
8:    uri: https://api.sandbox.braintreegateway.com:443/merchants/j9gwdfjdkxymhdgr/transactions
50:    uri: https://api.sandbox.braintreegateway.com:443/merchants/j9gwdfjdkxymhdgr/transactions/gasm2aq5/submit_for_settlement
93:    uri: https://api.sandbox.braintreegateway.com:443/merchants/j9gwdfjdkxymhdgr/transactions/gasm2aq5/void
138:    uri: https://api.sandbox.braintreegateway.com:443/merchants/j9gwdfjdkxymhdgr/transactions
180:    uri: https://api.sandbox.braintreegateway.com:443/merchants/j9gwdfjdkxymhdgr/transactions/hy1g7153/submit_for_settlement
222:    uri: https://api.sandbox.braintreegateway.com:443/merchants/j9gwdfjdkxymhdgr/transactions/hy1g7153
264:    uri: https://api.sandbox.braintreegateway.com:443/merchants/j9gwdfjdkxymhdgr/transactions/hy1g7153/void
309:    uri: https://api.sandbox.braintreegateway.com:443/merchants/j9gwdfjdkxymhdgr/transactions
351:    uri: https://api.sandbox.braintreegateway.com:443/merchants/j9gwdfjdkxymhdgr/transactions/dcjbnhzk/submit_for_settlement
394:    uri: https://api.sandbox.braintreegateway.com:443/merchants/j9gwdfjdkxymhdgr/transactions/dcjbnhzk/void
439:    uri: https://api.sandbox.braintreegateway.com:443/merchants/j9gwdfjdkxymhdgr/transactions
481:    uri: https://api.sandbox.braintreegateway.com:443/merchants/j9gwdfjdkxymhdgr/transactions/bsafgvhn/submit_for_settlement
525:    uri: https://api.sandbox.braintreegateway.com:443/merchants/j9gwdfjdkxymhdgr/transactions/bsafgvhn/void
567:    uri: https://api.sandbox.braintreegateway.com:443/merchants/j9gwdfjdkxymhdgr/customers
589:    uri: https://api.sandbox.braintreegateway.com:443/merchants/j9gwdfjdkxymhdgr/payment_methods
621:    uri: https://api.sandbox.braintreegateway.com:443/merchants/j9gwdfjdkxymhdgr/transactions
664:    uri: https://api.sandbox.braintreegateway.com:443/merchants/j9gwdfjdkxymhdgr/customers
686:    uri: https://api.sandbox.braintreegateway.com:443/merchants/j9gwdfjdkxymhdgr/payment_methods
716:    uri: https://api.sandbox.braintreegateway.com:443/merchants/j9gwdfjdkxymhdgr/payment_methods
748:    uri: https://api.sandbox.braintreegateway.com:443/merchants/j9gwdfjdkxymhdgr/transactions
790:    uri: https://api.sandbox.braintreegateway.com:443/merchants/j9gwdfjdkxymhdgr/transactions/722684ny/void
834:    uri: https://api.sandbox.braintreegateway.com:443/merchants/j9gwdfjdkxymhdgr/transactions
876:    uri: https://api.sandbox.braintreegateway.com:443/merchants/j9gwdfjdkxymhdgr/transactions/b25wkwrs/void
920:    uri: https://api.sandbox.braintreegateway.com:443/merchants/j9gwdfjdkxymhdgr/transactions
962:    uri: https://api.sandbox.braintreegateway.com:443/merchants/j9gwdfjdkxymhdgr/transactions/mnn5g3cd/submit_for_settlement
1006:    uri: https://api.sandbox.braintreegateway.com:443/merchants/j9gwdfjdkxymhdgr/transactions
1048:    uri: https://api.sandbox.braintreegateway.com:443/merchants/j9gwdfjdkxymhdgr/transactions/8fw0kke8/submit_for_settlement
1092:    uri: https://api.sandbox.braintreegateway.com:443/merchants/j9gwdfjdkxymhdgr/transactions
1134:    uri: https://api.sandbox.braintreegateway.com:443/merchants/j9gwdfjdkxymhdgr/transactions/0fmkw38j/submit_for_settlement
1178:    uri: https://api.sandbox.braintreegateway.com:443/merchants/j9gwdfjdkxymhdgr/transactions
1220:    uri: https://api.sandbox.braintreegateway.com:443/merchants/j9gwdfjdkxymhdgr/transactions/2wge34xf/submit_for_settlement
1262:    uri: https://api.sandbox.braintreegateway.com:443/merchants/j9gwdfjdkxymhdgr/transactions/advanced_search_ids
```

**Comment by whit537** [comment]
Eight out of 14 fixture files have Braintree API calls in them.

**Comment by whit537** [comment]
`TestPages.yml` is the only file that mixes Braintree with other API requests.

**Comment by whit537** [comment]
Alright, I think we should re-record all of the files that have Braintree in them.

**Comment by whit537** [comment]
That should get me back on track to record new tests under https://github.com/gratipay/gratipay.com/pull/4613.

**Comment by whit537** [comment]
```
$ gs
 M tests/py/fixtures/TestPages.yml
?? braintree.log
$ rm `cat braintree.log `
$ gs
 D tests/py/fixtures/BillingHarness.yml
 D tests/py/fixtures/TestHistory.yml
 D tests/py/fixtures/TestNMassPays.yml
 M tests/py/fixtures/TestPages.yml
 D tests/py/fixtures/TestPayday.yml
 D tests/py/fixtures/TestPayin.yml
 D tests/py/fixtures/TestRoutes.yml
 D tests/py/fixtures/TestsWithBillingHarness.yml
?? braintree.log
$ rm braintree.log 
$ ack braintree tests/py/fixtures/
$
```

**Comment by whit537** [comment]
Okay! I hand-edited `TestPages.yml` to remove the Braintree calls. So they're all gone from the fixtures now, and a re-run should repopulate them all. Lessee!

**Comment by whit537** [comment]
Hmmm ... 🤔 

```
tests/py/test_pages.py ...............F..........F.F.................
```

**Comment by whit537** [comment]
While I wait ... I went through our production processing settings (and then logged out of production again). We actually do have duplicate checking turned on there as well. The only other discrepancy is we don't have PayPal turned on in production, so I turned that off in the sandbox as well.

**Comment by whit537** [comment]
I think the failures are because the cassette exists. Isn't there a mode that adds new tests to an existing cassette?

http://vcrpy.readthedocs.io/en/latest/usage.html#record-modes

**Comment by whit537** [comment]
Let's try `new_episodes`?

**Comment by whit537** [comment]
Yeah, that's it. Okay!

```
CannotOverwriteExistingCassetteException: No match for the request (<Request (POST) https://api.sandbox.braintreegateway.com:443/merchants/bk8h97tqzyqjhtfn/customers>) was found. Can't overwrite existing cassette (u'/Users/whit537/personal/gratipay/gratipay.com/tests/py/fixtures/TestPages.yml') in your current record mode (u'once').
============================= 3 failed, 992 passed, 34 xfailed in 413.22 seconds =============================

```

**Comment by whit537** [comment]
```
========================================= 46 passed in 41.22 seconds =========================================
$ gs
 M tests/py/fixtures/TestPages.yml
?? tests/py/fixtures/BillingHarness.yml
?? tests/py/fixtures/TestHistory.yml
?? tests/py/fixtures/TestNMassPays.yml
?? tests/py/fixtures/TestPayday.yml
?? tests/py/fixtures/TestPayin.yml
?? tests/py/fixtures/TestRoutes.yml
?? tests/py/fixtures/TestsWithBillingHarness.yml
$
```

**Comment by whit537** [comment]
Moving to PR in https://github.com/gratipay/gratipay.com/pull/4616 ...

**Comment by whit537** [comment]
Back to resolve Travis [issue](https://github.com/gratipay/gratipay.com/pull/4616#issuecomment-328326635) ...

> Two of the tests are trying to add calls to the cassette.
>
>`test_on_the_fence_can_browse`
>`test_new_participant_can_browse `

**Comment by whit537** [comment]
What's the diff between local and CI envs that is causing this?

**Comment by whit537** [comment]
The `_browse` tests do a lot ... notably the `anon` case is fine.

**Comment by whit537** [comment]
`on_the_fence` is [failing early](https://travis-ci.org/gratipay/gratipay.com/builds/273746696#L1503) in `browse` when we try to pre-pop the receipt page.

`new_participant` is [failing](https://travis-ci.org/gratipay/gratipay.com/builds/273746696#L1440) in `routes/credit-card.spt`.

**Comment by whit537** [comment]
Hmm ... each fails locally when run individually ... both on `routes/credit-card.spt` 🤔

**Comment by whit537** [comment]
Run together the two fail as at Travis.

```
$ ARGS="tests/py/test_pages.py" t
```

**Comment by whit537** [comment]
Seems easy enough to record and add, but why didn't they get added in the first big re-record?

**Comment by whit537** [comment]
Re-run of all of `tests/py/` fails in the same way.

**Comment by whit537** [comment]
Repurposing this ticket to be my verbose log for shipping https://github.com/gratipay/gratipay.com/pull/4598 in the next ... three hours?

**Comment by whit537** [comment]
Reviewing work to be done ...

**Comment by whit537** [comment]
Okay! https://github.com/gratipay/gratipay.com/pull/4616 is green! I'm going to merge #4614 and #4616, and then rebase the #4598 chain.

**Comment by whit537** [comment]
Done. Almost done organizing tickets ...

**Comment by whit537** [comment]
Alright, so that's four more PRs needed against #4598:

- error checking
- optional
- receipt
- ToS

Okay! 👊 👊 🍖

**Comment by whit537** [comment]
Debating whether to self-review-and-merge the four ready PRs on #4598 or start into new work and then review/merge all eight at once. Eight sounds like a lot, thinking let's get these four in now.

**Comment by whit537** [comment]
I mean, Travis will take half an hour to run anyway.

**Comment by whit537** [comment]
Which actually raises a good point: optimization is to merge the last in the chain that is ready to merge, to avoid the rebase-and-wait-for-Travis dance overhead.

**Comment by whit537** [comment]
Re: dance overhead: https://www.youtube.com/watch?v=ovo6zwv6DX4.

**Comment by whit537** [comment]
Alright, 4-chain updated, go Travis go! I know there are a couple things to fix on https://github.com/gratipay/gratipay.com/pull/4617, what about the others?

**Comment by whit537** [comment]
#4606 is fine, #4604 is what #4617 exists to fix. Knocking out #4617 issues ...

**Comment by whit537** [comment]
#4606 is green but #4604 ain't. What gives?

**Comment by whit537** [comment]
Probably will need to bump #4609 back off in the interest of time.

**Comment by whit537** [comment]
Errors, receipt, terms. Next chain!

**Comment by whit537** [comment]
Sup with https://github.com/gratipay/gratipay.com/pull/4604 tho?

**Comment by whit537** [comment]
Okay, fixed.

I can do a simpler optional rationalization (no folding, just the two sections) on #4613.

**Comment by whit537** [comment]
#4604 is green, almost have #4617 ready.

**Comment by whit537** [comment]
Alright, #4617 over to Travis, hitting #4613.

**Comment by whit537** [comment]
#4617 is green! I am modifying/extending functionality on #4613.

**Comment by whit537** [comment]
[slack](https://gratipay.slackarchive.io/gratipay/page-100/ts-1505044170000042)

**Comment by whit537** [comment]
Alright, squeezing out last little bit of CI floofle on #4621 and then the new three-chain is ready for final review and merge. 👍

**Comment by whit537** [comment]
Landed #4598!
